src/chk/chk_engine.c

```diff
@@ -963,6 +971,184 @@ chk_engine_cont_list_remote_cb(void *args, uint32_t rank, uint32_t phase,
 	return rc;
 }
 
+static int
+chk_engine_cont_orphan(struct chk_pool_rec *cpr, struct chk_cont_rec *ccr, struct cont_svc *svc)
+{
+	struct chk_instance		*ins = cpr->cpr_ins;
+	struct chk_property		*prop = &ins->ci_prop;
+	struct chk_bookmark		*cbk = &cpr->cpr_bk;
+	struct chk_report_unit		 cru = { 0 };
+	Chk__CheckInconsistClass	 cla;
+	Chk__CheckInconsistAction	 act;
+	char				 msg[160] = { 0 };
+	uint32_t			 options[2];
+	uint32_t			 option_nr = 0;
+	int				 decision = -1;
+	int				 result = 0;
+	int				 rc = 0;
+
+	cla = CHK__CHECK_INCONSIST_CLASS__CIC_CONT_NONEXIST_ON_PS;
+	act = prop->cp_policies[cla];
+	cbk->cb_statistics.cs_total++;
+
+	switch (act) {
+	case CHK__CHECK_INCONSIST_ACTION__CIA_DEFAULT:
+		/*
+		 * If the container is not registered to the container service, then destroy the
+		 * orphan container to release space by default.
+		 *
+		 * XXX: Currently, we do not support to add the orphan container back to the CS,
+		 *	that may be implemented in the future when we have enough information to
+		 *	recover necessary prop/attr for the orphan container.
+		 *
+		 * Fall through.
+		 */
+	case CHK__CHECK_INCONSIST_ACTION__CIA_TRUST_PS:
+		/* Fall through. */
+	case CHK__CHECK_INCONSIST_ACTION__CIA_DISCARD:
+		if (prop->cp_flags & CHK__CHECK_FLAG__CF_DRYRUN) {
+			cbk->cb_statistics.cs_repaired++;
+		} else {
+			result = ds_cont_destroy_orphan(svc, ccr->ccr_uuid);
+			if (result != 0)
+				cbk->cb_statistics.cs_failed++;
+			else
+				cbk->cb_statistics.cs_repaired++;
+		}
+		break;
+	case CHK__CHECK_INCONSIST_ACTION__CIA_IGNORE:
+		/* Report the inconsistency without repair. */
+		cbk->cb_statistics.cs_ignored++;
+		break;
+	default:
+		/*
+		 * If the specified action is not applicable to the inconsistency,
+		 * then switch to interaction mode for the decision from admin.
+		 *
+		 * Fall through.
+		 */
+	case CHK__CHECK_INCONSIST_ACTION__CIA_INTERACT:
+		options[0] = CHK__CHECK_INCONSIST_ACTION__CIA_DISCARD;
+		options[1] = CHK__CHECK_INCONSIST_ACTION__CIA_IGNORE;
+		option_nr = 2;
+		break;
+	}
+
+report:
+	cru.cru_gen = cbk->cb_gen;
+	cru.cru_cla = cla;
+	cru.cru_act = option_nr != 0 ? CHK__CHECK_INCONSIST_ACTION__CIA_INTERACT : act;
+	cru.cru_rank = dss_self_rank();
+	cru.cru_option_nr = option_nr;
+	cru.cru_pool = (uuid_t *)&cpr->cpr_uuid;
+	cru.cru_cont = (uuid_t *)&ccr->ccr_uuid;
+	snprintf(msg, 159, "Check engine detects orphan container "DF_UUIDF"/"DF_UUIDF,
```

Comment:
[minor ]use a Macro like MAX_MSG_LEN - 1  looks better

Reply:
> [minor ]use a Macro like MAX_MSG_LEN - 1 looks better

Agree. But currently because different report events use different msg lengths, then we do not have a fixed MAX_MSG_LEN definition. I will cleanup and unify them in subsequent patches.